### PR TITLE
intercept browser's failure (issue #448)

### PR DESCRIPTION
**Pull request by Arlon1**
When opening a link the browser may return a value unequal to zero.
This gets recognized and then printed to stderr.

If this happens during the "open-in-browser-and-mark-read"-operation neither
the article in question is marked as read nor it is scrolled down.

Unfortunately I had to change the method declaration of open_in_browser from void
to bool in order to transmit one bit of information to the outside of that
particular function. Although this should not be an issue as this is
syntactically allowed similar a line of code only persisting of a literal.

### CODE REVIEW

File: src/view.cpp [hunk 1]
```diff
@@ -359,8 +359,16 @@ void view::open_in_browser(const std::string& url) {
...
-	utils::run_interactively(cmdline, "view::open_in_browser");
+	// store the shell return value
+	int status_browser = utils::run_interactively(cmdline, "view::open_in_browser");
+	// the browser returned a value unequal to zero
+	if (status_browser) {
+		std::clog << strprintf::fmt(_("Starting the browser failed with error code %d"), status_browser) << std::endl;
+	}
 	pop_current_formaction();
+
+	if (status_browser) return false;
```
Comment (Minoru) [inline src/view.cpp]: `return status_browser == 0;` is simpler yet just as readable.

File: src/view.cpp [hunk 2]
```diff
@@ -359,8 +359,16 @@ void view::open_in_browser(const std::string& url) {
 		cmdline.append("'");
 	}
 	stfl::reset();
-	utils::run_interactively(cmdline, "view::open_in_browser");
+	// store the shell return value
```
Comment (Minoru) [inline src/view.cpp]: You don't have to write comments that explain basic things like assignment operator. Reserve them for tricky stuff (basically anything that took you a while to think up) and explanations of your motivation (it's easy enough to figure out *what* the code does, but it's much harder to find out *why* it does that).

File: src/view.cpp [hunk 3]
```diff
@@ -359,8 +359,16 @@ void view::open_in_browser(const std::string& url) {
...
+	int status_browser = utils::run_interactively(cmdline, "view::open_in_browser");
```
Comment (Minoru) [inline src/view.cpp]: "Exit code" is a common term for the value that a program returns. Please use it instead of "status" here.

File: src/view.cpp [hunk 4]
```diff
@@ -359,8 +359,16 @@ void view::open_in_browser(const std::string& url) {
...
+	// the browser returned a value unequal to zero
+	if (status_browser) {
+		std::clog << strprintf::fmt(_("Starting the browser failed with error code %d"), status_browser) << std::endl;
```
Comment (Minoru) [inline src/view.cpp]: Newsbeuter has custom logging facility; grep sources for `LOG` macro. I think it's useful to use `level::DEBUG` here, and show the `cmdline` in addition to its exit code.

File: src/itemlist_formaction.cpp [hunk 5]
```diff
@@ -84,11 +84,19 @@ void itemlist_formaction::process_operation(operation op, bool automatic, std::v
...
-				visible_items[itempos].first->set_unread(false);
-				v->get_ctrl()->mark_article_read(visible_items[itempos].first->guid(), true);
-				v->open_in_browser(visible_items[itempos].first->link());
+				// store if open_in_browser() succeded
+				bool browser_success = v->open_in_browser(visible_items[itempos].first->link());
+				// do only mark as unread if open_in_browser() succeded
+				if (browser_success) {
+					visible_items[itempos].first->set_unread(false);
+					v->get_ctrl()->mark_article_read(visible_items[itempos].first->guid(), true);
+				}
```
Comment (Minoru) [inline src/itemlist_formaction.cpp]: This code has two problems:
1. no error is displayed if browser fails;
2. it carries on to `OP_NEXTUNREAD` regardless of browser's exit code.

I think that instead of handling the happy case (`if (browser_success) …`), this should handle the failure (`if (! browser_success) …`). it could just show an error (grep for `show_error`) and `break` from this whole `switch` (or maybe `return` from the function; you decide).

This also means you won't have to add `browser_success` to the condition below.

File: src/itemlist_formaction.cpp [hunk 6]
```diff
@@ -84,9 +84,14 @@ void itemlist_formaction::process_operation(operation op, bool automatic, std::v
 		LOG(level::INFO, "itemlist_formaction: opening item at pos `%s'", itemposname);
 		if (itemposname.length() > 0 && visible_items.size() != 0) {
 			if (itempos < visible_items.size()) {
+
+				bool browser_success = v->open_in_browser(visible_items[itempos].first->link());
+				if (!browser_success)
```
Comment (Minoru) [inline src/itemlist_formaction.cpp]: This still doesn't display the error to the user. Do you need any help with this?

File: src/view.cpp [hunk 7]
```diff
@@ -359,8 +359,14 @@ void view::open_in_browser(const std::string& url) {
 		cmdline.append("'");
 	}
 	stfl::reset();
-	utils::run_interactively(cmdline, "view::open_in_browser");
+	int browser_exit_code = utils::run_interactively(cmdline, "view::open_in_browser");
+	if (browser_exit_code == 0) {
```
Comment (Minoru) [inline src/view.cpp]: A typo, should be `!=` because zero exit code means *success*.

File: src/view.cpp [hunk 8]
```diff
@@ -359,8 +359,14 @@ void view::open_in_browser(const std::string& url) {
...
+		LOG(level::DEBUG, "Starting the browser failed with error code %d - command was %s\n", browser_exit_code, cmdline);
```
Comment (Minoru) [inline src/view.cpp]: 1. No need to include "\n" at the end of the message—the logger will do that for you automatically.
2. We usually prefix our log lines with the name of the function to make it easier to follow; in this case it should be `utils::run_interactively`.

File: src/view.cpp [hunk 9]
```diff
@@ -359,8 +359,14 @@ void view::open_in_browser(const std::string& url) {
...
+	}
 	pop_current_formaction();
+
+	if (browser_exit_code) return false;
```
Comment (Minoru) [inline src/view.cpp]: You missed my previous comment on this: `return browser_exit_code == 0` is easier to understand. Just imagine the thought process of someone who will read the current version:

1. Ok, so `browser_exit_code` is an `int`, and it'll be converted to `bool` here.
2. So if it's non-zero, the condition evaluates to `true`, we're going to the first branch and function returns `false`;
3. ...And if it's zero, the condition is `false` and we return `true`.
4. Ah, so it just checks if the exit code is zero.

If you write `return browser_exit_code == 0`, you're essentially eliminating the first three steps above and jump right to the conclusion.

**Comment by Arlon1** [comment]
Thanks for your detailed advice! I really appreciate it.
But there's one thing to mention: In ```itemlist_formaction::process_operation()``` no error message is printed because that happens in ```view::open_in_browser()``` (where you mentioned the call of the ```LOG()``` function.
I thought this might be a better better place for it because it happens straight after ```system()``` and ```open_in_browser()``` has only to return a bool. Where do you think should this error message belong to?

**Comment by Minoru** [comment]
> In itemlist_formaction::process_operation() no error message is printed because that happens in view::open_in_browser() (where you mentioned the call of the LOG() function.

I was asking for a `show_error` in `process_operation` and for `LOG` in `open_in_browser`. Sorry, I should've explained the different immediately.

The calls to `LOG` write the error to the log file, which users can explicitly request using `--log-file` option. Messages with `level::USERERROR` are also written to the error-log (configured by `error-log` command in the config). Finally, the calls to `view::show_error` display a message at the bottom of the screen.

We're using `LOG` with `level::USERERROR` for errors that users *should* be aware of, like dead RSS feeds. Other log levels are used for debugging (I often ask users to run Newsbeuter with logging enabled to figure out where the problem is). The `show_error` is used for errors that users *must* be aware of.

In short, we're using `LOG` (mostly) for debugging purposes, and `show_error` to notify users of anything.

In case of this PR, we should notify users about browser's non-zero exit code. There's no need to go into detail like exit codes and commands—it's sufficient to say something like "browser failed to open the link".

File: src/itemlist_formaction.cpp [hunk 10]
```diff
@@ -86,8 +86,10 @@ void itemlist_formaction::process_operation(operation op, bool automatic, std::v
 			if (itempos < visible_items.size()) {
 
 				bool browser_success = v->open_in_browser(visible_items[itempos].first->link());
-				if (!browser_success)
+				if (!browser_success) {
+					v->show_error(_("browser failed to open the link"));
```
Comment (Minoru) [inline src/itemlist_formaction.cpp]: I'm sorry to cavil on such a minor detail, but it's user-facing and thus quite important: can you please make this message a whole sentence, with capital "b" at the beginning and a period at the end? All the other call to `show_error` do that.

**Comment by Minoru** [comment]
A-and… merged!! Thank you very much for working on this!

If you're looking for the next issue to tackle, #449 might be a good choice since you already got the context.
